File: Java/0050_Pow_x_n.java
// 2020-08-04 22:04:22

class Solution {
    public double myPow(double x, int n) {
        double result = 1;
        long N = Math.abs((long) n);

        while (N != 0) {
            if ((N & 1) == 1)
                result *= x;
            x *= x;
            N >>= 1;
        }

        return n < 0 ? 1 / result : result;
    }
}

File: C++/0050_powxn.cc
// 2020-08-04 23:44:38

#include <cstdlib>

class Solution
{
public:
    double myPow(double x, int n)
    {
        long N = abs(n);
        double ret = 1;
        while (N != 0)
        {
            if ((N & 1) == 1)
                ret *= x;
            x *= x;
            N >>= 1;
        }
        return n < 0 ? 1 / ret : ret;
    }
};
